### PR TITLE
Fix how we cancel the context in the builtin backup engine

### PR DESCRIPTION
**Pull request by frouioui**
## Description

This PR makes sure we cancel the context only at the correct time. It was getting canceled too early which could lead to incomplete/failing backups with storage implementations that write concurrently (Ceph and S3).

This regression was introduced by: https://github.com/vitessio/vitess/pull/16856, backporting this to release-21.0 as the issue was introduced there. It would lead to the following error:

```
commerce/0 (zone1-0000000102): time:{seconds:1732657021 nanoseconds:695573000} file:"builtinbackupengine.go" line:509 value:"resetting mysqld super_read_only to true"
commerce/0 (zone1-0000000102): time:{seconds:1732657021 nanoseconds:696651000} file:"builtinbackupengine.go" line:526 value:"restarting mysql replication"
E1126 15:37:02.757871   98619 main.go:56] rpc error: code = Unknown desc = TabletManager.Backup on zone1-0000000102: operation error S3: PutObject, https response error StatusCode: 0, RequestID: , HostID: , canceled, context canceled;operation error S3: PutObject, https response error StatusCode: 0, RequestID: , HostID: , canceled, context canceled;operation error S3: PutObject
```

### CODE REVIEW

**Review by rohit-nayak-ps — APPROVED** [review]

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 1]
```diff
@@ -604,7 +604,15 @@ func (be *BuiltinBackupEngine) backupFiles(
...
-	defer cancel()
+	defer func() {
+		// We may still have operations in flight that require a valid context, such as adding files to S3.
+		// Unless we encountered an error, we should not cancel the context, this is taken care of later
+		// in the process. If we encountered an error however, we can safely cancel the context as we should
+		// no longer work on anything and exit fast.
+		if finalErr != nil {
+			cancel()
+		}
+	}()
```
Comment (shlomi-noach) [inline go/vt/mysqlctl/builtinbackupengine.go:615]: I'm not sure I follow:
- It looks as if the function can exit and still leave the context active, in which case, what is cancelling it?
- If we still have operations in flight, why would we exit the function? Should we not wait until everything is complete?
Something feels off here, as an anti-pattern.

**Review by shlomi-noach — COMMENTED** [review]
Please see inline question - I unfortunately do not follow the logic.

File: go/vt/mysqlctl/builtinbackupengine.go (hunk 1 above)
Comment (frouioui) [reply go/vt/mysqlctl/builtinbackupengine.go:615]: > what is cancelling it?

The caller of ExecuteBackup will eventually cancel it, whether it is through a gRPC call or through vtbackup, the context always gets canceled.

Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go:615]: Thank you for the clarity! ❤️

**Review by shlomi-noach — APPROVED** [review]

**Pull request merged**